types/ckeditor__ckeditor5-engine/src/conversion/upcastdispatcher.d.ts

```diff
@@ -46,33 +50,60 @@ export interface UpcastConversionApi {
         position: Position;
         cursorParent: Element;
     };
-    updateConversionResult(element: Element, data: UpcastConversionData, conversionApi: UpcastConversionApi): void;
+    updateConversionResult(element: Element, data: UpcastConversionData): void;
 }
 
-export default class UpcastDispatcher implements Emitter {
-    conversionApi: UpcastConversionApi;
+export interface UpcastEventDataTypes {
+    element: UpcastConversionData<ViewElement>;
+    text: UpcastConversionData<ViewText>;
+    documentFragment: UpcastConversionData<ViewDocumentFragment>;
+}
+
+export type UpcastEventArgs<K extends string = string> = K extends keyof UpcastEventDataTypes
+    ? [UpcastEventDataTypes[K], UpcastConversionApi]
+    : K extends "viewCleanup"
+    ? [ViewDocumentFragment | ViewElement]
+    : K extends `element:${infer N}`
+    ? [UpcastConversionData<ViewElement & { name: N; }>, UpcastConversionApi]
+    : K extends `${infer NS}:${string}`
+    ? NS extends keyof UpcastEventDataTypes
+        ? [UpcastEventDataTypes[NS], UpcastConversionApi]
+        : any[]
+    : any[];
+
+export type UpcastDispatcherCallback<N extends string, S extends Emitter = Emitter> = (
+    info: EventInfo<S, N>,
+    ...args: UpcastEventArgs<N>
+) => void;
+
+export default class UpcastDispatcher {
```

Comment:
I have been out of touch with this project lately, so I must ask: Is the interface `Emitter` so broken that you omitted the `implements Emitter`? Asking cause even though you added all the required methods, it would be difficult to track those classes that implement it in case the original `Emitter` changes.

Reply:
Yes, essentially. This class is compatible with the Emitter interface in #54986, but `Emitter["on"]` without your PR is currently defined as

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/16c18e96e074c827cc01d3d20beeecae3df970cf/types/ckeditor__ckeditor5-utils/src/emittermixin.d.ts#L15-L19

And since data here is not `DomEventData`, I cannot have `implements Emitter`